[Security] New Password Policy listener

Spomky opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | yes <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | none
| License       | MIT
| Doc PR        | symfony/symfony-docs#... [**TODO**]

This PR adds the possibility to define security policy on the password.
When at least one rule is defined, the PasswordPolicyListener will be called to verify the submitted password (on login) is acceptable.
The aim of this feature is to enforce the update of passwords.
Hereafter some use cases:
* There were no rules when designing the app. Now password length should be at least 8 characters and short passwords should be changed.
* The minimum password strength was 2 from a scale from 0 to 5. The IT Service decided to increase the minimum strength to 2.
* The user database has leaked and is available on the internet. Leaked passwords are now prohibited.

The listener has a high priority (512, TBD) in order to be called before the user is fetched and the password is checked against the value in the database. The policy is only based on the plain text password value and not on the user state (e.g. last password update).

Todo:
* [x] Tests
* [ ] Documentation

wouterj commented:
Hi! Thanks for the series of security feature PRs!

However, I have some doubts about this particular feature:

A) If we ignore length constraints, all other constraints seem more suited in a registration process and not an authentication process. I can't recall seeing a website that tells me I have an unsafe password during authentication. For the basic constraints like length, I don't think we need to integrate with the Validator component (we already check on length btw)

B) The badges and passport listeners are created to centralize security sensitive code. I don't consider this validator a security critical feature. Also, given the validator component has its own simple high level API, I think using it directly in your own listener or authenticator instead of this badge is more readable.

Spomky commented:
Hi @wouterj,

Many thanks for your feedback and your advice.
The main idea is that I wanted to prevent login with a compromised password. At first sight, I duplicated the logic from the `NotCompromised` constraint, but after a short discussion with @nicolas-grekas we went to the idea of a badge that could verify the input against any other constraint.

nicolas-grekas commented:
@wouterj WDYT of our responses to your comment?

wouterj commented:
Sorry, I have been thinking about this PR the past weeks, but I didn't write down my thoughts here.

I think the "prevent login with a compromised password" idea is very interesting. This is becoming a more common practice, and is also recommended by NIST. However, I think I would like a very scoped badge rather than a generic badge like this one, to "limit" the usages, which makes it easier to make changes on later on.

Two other things I have in mind are:
* We were cautious with storing plaintext passwords in badges (maybe overly so). E.g. the `PasswordCredentials` automatically clears the plaintext password after it is checked. So I'm a bit apprehensive with introducing another badge containing the plaintext password (I realize we did add an `PasswordUpgradeBadge` before though).
* I'm not that fond of configuring a "password policy" (rules to determine if a user should change their password) in an authenticator (this also applies to `PasswordUpgradeBadge`). I would rather have the authenticator only add a `PasswordCredentials` and then have a `PasswordPolicyListener` that checks whether the password passes the policy or not. We might even add this as a new configuration option `security.password_policy` or the like.

chalasr commented:
I agree such concern would be better handled outside of the authenticator despite the "opportunistic upon login" aspect.

wouterj commented:
Just a dummy example, we might do something like this:

```php
interface PasswordPolicy
{
    public function accept(string $plaintextPassword): bool;
}
```
```yaml
security:
    password_policy:
        - NotCompromisedPolicy
        - 'App\Security\CustomPolicy'
```

A listener on `CheckPassportEvent` would then check a `PasswordCredentials` with these policies and block authentication before the `CheckCredentialsListener` is executed.

We could extract the haveibeenpwned integration from the Validation component (into PasswordHasher?) and reuse it in both the constraint and policy. Maybe we implement some more policies, but we should make sure to only implement policies following best practices (e.g. from NIST). This last point is also why I'm not keen to integrate with the Validator component, this makes it way to easy to follow bad practices imho.

This would make it quite easy to enforce these policies, while still allowing users to implement not-best practice policies when necessary (e.g. a client of us requires us to have an "update password every x months" policy).

Spomky commented:
Many thanks for your feedback.

>  However, I think I would like a very scoped badge rather than a generic badge like this one, to "limit" the usages, which makes it easier to make changes on later on and discourage bad practices (e.g. requiring a password to contain at least 1 special character, etc.).

To be honest, this is what I usually do and that was used in recent workshops ([badge](https://github.com/Spomky-Labs/meteofony/blob/sessions/2023-03-22/src/Security/PasswordNotCompromisedBadge.php) and [listener](https://github.com/Spomky-Labs/meteofony/blob/sessions/2023-03-22/src/Security/PasswordNotCompromisedBadgeListener.php)). It works fine and most of the code is mostly copied from the constraint.
If you are ok with copy/paste, I will change it in a dedicated badge

> We were cautious with storing plaintext passwords in badges (maybe overly so)

OK noted. When the badge is marked as resolved, I will clear it.

>  I would rather have the authenticator only add a PasswordCredentials and then have a PasswordPolicyListener 

OK. To me, this means this `PasswordPolicyListener` will be executed before any other listener and especially the PasswordCredentialsListener otherwise the password is cleared.

Spomky commented:
> Just a dummy example, we might do something like this:

I understand the logic. Let me time for implementing it.

wouterj commented:
> I understand the logic. Let me time for implementing it.

Cool, thanks for continueing this. Take it easy, the next future freeze is far away 🙂 and feel free to ping me if you need help getting this ready for 6.4.

Spomky commented:
> Cool, thanks for continueing this. Take it easy, the next future freeze is far away 🙂 and feel free to ping me if you need help getting this ready for 6.4.

🫢I took time to redo this PR (only the component). I concur with you, the design is much better now.

stof commented on src/Symfony/Component/Security/Http/EventListener/PasswordPolicyListener.php:
```diff
@@ -0,0 +1,54 @@
+     */
+    private array $policies = [];
+
+    public function add(PolicyInterface $policy): void
```
you should inject an iterable into the constructor instead, which would allow the bundle to configure that using an IteratorArgument that lazy-loads the services

stof commented on src/Symfony/Bundle/SecurityBundle/Tests/Functional/app/PasswordPolicy/BlacklistPasswordPolicy.php:
```diff
@@ -0,0 +1,16 @@
+
+use Symfony\Component\Security\Http\Authenticator\Passport\Policy\PolicyInterface;
+
+final class BlacklistPasswordPolicy implements PolicyInterface
```
You should name it `Blocklist` rather than `Blacklist`

stof commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php:
```diff
@@ -190,6 +190,13 @@ public function load(array $configs, ContainerBuilder $container)
         $container->getDefinition('security.authorization_checker')->setArgument(2, false);
         $container->getDefinition('security.authorization_checker')->setArgument(3, false);
+
+        if ($container->hasDefinition('security.password_policy_listener')) {
```
why is this conditional ? The service is always defined.

stof commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php:
```diff
@@ -190,6 +190,13 @@ public function load(array $configs, ContainerBuilder $container)
         $container->getDefinition('security.authorization_checker')->setArgument(3, false);
+
+        if ($container->hasDefinition('security.password_policy_listener')) {
+            $definition = $container->getDefinition('security.password_policy_listener');
```
Should we remove the service definition in case we don't have any policies ?

stof reviewed (commented):
the XSD file must be updated

Spomky commented:
Many thanks @stof. I addressed your comments.

I have an issue with the test that fails. It looks like the value of `$config['password_policy']` is always an empty list. It is correctly populated in `$configs`, but seems to be replaced by the configuration processor call.
@wouterj suggested to change `->fixXmlConfig('password_policy', 'password_policy')` in `MainConfiguration` but the result is always the same.
Do you have any clue on what is causing this issue? I certainly missed something stupid, but I cannot figure out what?

stof commented:
Well, I don't think using the same name for the singular and the plural works fine. I think you should use an actual plural name `password_policies`, which will solve this issue of using a single and plural with the same value.

Spomky commented:
That was my assumption, unfortunately the service list is still empty

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php:
```diff
@@ -190,6 +190,13 @@ public function load(array $configs, ContainerBuilder $container)
+        if (!$config['password_policies']) {
+            $container->removeDefinition('security.password_policy_listener');
+        } else {
+            $policyServices = array_map(static fn (string $id) => new Definition($id), $config['password_policies']);
```
I don't think this is correct: entries in the list of policies cannot directly map to a service definition. that'd mean not being allowed to configure their constructor arguments for example.
Instead, we could use `new Reference`, but I'm wondering if we're not missing a DI tag to do the link, and use `tagged_iterator()`

Spomky replied:
Correct. I modified the extension and my service definitions in the test.
The use of a tag is a good idea, but can we detect if a service is tagged in this method?

stof replied:
not in this method. We would have to do it in a compiler pass (or use `tagged_iterator` without checking things)

nicolas-grekas commented on src/Symfony/Component/Security/Http/Authenticator/Passport/Policy/PolicyInterface.php:
```diff
@@ -0,0 +1,22 @@
+
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+
+interface PolicyInterface
```
I'd name this one PasswordPolicyInterface, that's what we use everywhere else in the PR

Spomky replied:
Correct. Done

nicolas-grekas commented on src/Symfony/Component/Security/Http/Authenticator/Passport/Policy/PolicyInterface.php:
```diff
@@ -0,0 +1,22 @@
+    /**
+     * @throws AuthenticationException
+     */
+    public function verify(#[\SensitiveParameter] string $plaintextPassword): void;
```
can't we make this return a bool and throw in the listener? That'd allow removing some boilerplate from implementations.

Spomky replied:
I changed the method signature and now it returns a `bool`. The exception is thrown by the listener.

yceruto commented on src/Symfony/Component/Security/Http/Authenticator/Passport/Policy/PasswordConstraintPolicy.php:
```diff
@@ -0,0 +1,38 @@
+     */
+    public function __construct(
+        private readonly ValidatorInterface $validator,
+        private readonly array $constraints = []
```
missing comma at the end

yceruto commented on src/Symfony/Component/Security/Http/Authenticator/Passport/Policy/PasswordPolicyInterface.php:
```diff
@@ -0,0 +1,22 @@
+    /**
+     * @throws AuthenticationException
+     */
+    public function verify(#[\SensitiveParameter] string $plaintextPassword): bool;
```
Could we stipulate that, instead of returning a bool, it must either throw a `PasswordPolicyException` or return void?

Spomky committed:
New Password Policy listener

Spomky commented:
Closing as the concept should be revised in depth.
